src/java/org/apache/cassandra/db/virtual/AccordVirtualTables.java

```diff
@@ -73,6 +90,88 @@ public DataSet data()
+    {
+        private CommandStoreCache(String keyspace)
+        {
+            super(parse(keyspace,
+                        "Accord Command Store Cache Metrics",
+                        "CREATE TABLE accord_command_store_cache(\n" +
+                        "  id int,\n" +
+                        "  queries bigint,\n" +
+                        "  hits bigint,\n" +
+                        "  misses bigint,\n" +
+                        "  PRIMARY KEY (id)" +
+                        ')'));
+        }
+
+        @Override
+        public DataSet data()
+        {
+            AccordService accord = (AccordService) AccordService.instance();
+            SimpleDataSet result = new SimpleDataSet(metadata());
+
+            Map<Integer, AccordStateCache.ImmutableStats> statsByStore = 
+                accord.node().commandStores().map(PreLoadContext.empty(),
+                                                  store -> ((AccordCommandStore) store.commandStore()).cache().stats());
+
+            for (Map.Entry<Integer, AccordStateCache.ImmutableStats> stats : statsByStore.entrySet())
+            {
+                result.row(stats.getKey());
+                result.column("queries", stats.getValue().queries);
+                result.column("hits", stats.getValue().hits);
+                result.column("misses", stats.getValue().misses);
+            }
+
+            return result;
+        }
+    }
+
+    public static final class MigrationStates extends AbstractVirtualTable
+    {
+        private MigrationStates(String keyspace)
+        {
+            super(parse(keyspace,
+                        "Accord Migration State",
+                        "CREATE TABLE accord_migration_state(\n" +
+                        "  keyspace_name text,\n" +
+                        "  table_name text,\n" +
+                        "  table_id uuid,\n" +
+                        "  target_protocol text,\n" +
+                        "  migrated_ranges text,\n" +
+                        "  migrating_ranges_by_epoch text,\n" +
+                        "  PRIMARY KEY (keyspace_name, table_name)" +
+                        ')'));
+        }
+
+        @Override
+        public DataSet data()
+        {
+            SimpleDataSet result = new SimpleDataSet(metadata());
+
+            ClusterMetadata cm = ClusterMetadata.current();
+            ConsensusMigrationState snapshot = cm.consensusMigrationState;
+
+            for (TableMigrationState state : snapshot.tableStates())
+            {
+                result.row(state.keyspaceName, state.tableName);
+                result.column("table_id", state.tableId.asUUID());
+                result.column("target_protocol", state.targetProtocol.toString());
+
+                List<String> primitiveMigratedRanges = state.migratedRanges.stream().map(Objects::toString).collect(toImmutableList());
+                result.column("migrated_ranges", PojoToString.pojoToString(primitiveMigratedRanges, "MINIFIED-JSON"));
```

Comment:
makes me wonder if `frozen<list<text>>` is better...  we wouldn't need to convert it at all and it would be easier for users to access in CQL?

Reply:
Yeah, this and the comment for `migrating_ranges_by_epoch` make sense. Let's try frozen collections instead...